### ISSUE TITLE
Time taken to query Cloudant local view vs Couch db view vs is large

### ISSUE THREAD

**Issue by gauravbh1** [cloudant/java-cloudant#410]
This issue is right now more of a query and might be converted into a bug based on the response that's provided by the development team.

I am using Cloudant client to query views in both Couch DB as well as Cloudant local.
I have the same view in both the dbs and same java code trying to read the response from the view using Cloudant client api.

The issue that I see is that time taken to get a response from a view in Cloudant local is far more than what it takes to query the same view in Couch DB.
e.g Using the code snippet given below, the time taken to get the response from Couch DB takes around 15-20 ms whereas time taken by the same code for Cloudant local is 215-225 ms.

My query is that is there any specific tuning required on Cloudant local to get the code working as efficiently as it works in Couch db? Or is there anything wrong with the way I am using the api?

-CloudantClient (java-cloudant) version(s) that are affected by this issue.
2.10.0
-Java version (including vendor and platform).
java version "1.8.0_141"
Java(TM) SE Runtime Environment (build 1.8.0_141-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.141-b15, mixed mode)

-A small code sample that demonstrates the issue.
View definition in Couch db (2.1.0)/Cloundant local.
```
function (doc) {
  if(doc.type == "test"){
    emit(doc._id, 1);
  }
}
```

Java code using cloudant java api
```
		CloudantClient client = null;
		try {
			client = ClientBuilder.url(new URL(COUCHDB_URL))
					.username(COUCH_USERNAME)
					.password(COUCH_PASSWORD)
					.build();
		} catch (MalformedURLException e) {

		}
		ViewResponse<ComplexKey, String> response = null;
		  Database db = client.database("mydb1", true);
		  if(db != null) {
		  	ViewRequestBuilder viewBuilder = db.getViewRequestBuilder("mydesigndoc", "testView");
		  	if(viewBuilder != null) {
		  		ViewRequest<ComplexKey, String> request = viewBuilder.newRequest(Key.Type.COMPLEX,String.class)
		  				.includeDocs(false)
		  				.build();

		  		try {
		  			long startmillis = System.currentTimeMillis(); 
		  			if(request != null) {						
		  				response = request.getResponse();
		  			}
		  			long endmillis = System.currentTimeMillis(); 
		  			
		  			long diff =  endmillis - startmillis;
		  			System.out.println("execution time is "+diff);
		  		} catch (IOException e) {
		  			e.printStackTrace();
		  		}
		  	}
		  }
```

**Comment by ricellis** [comment]
From a java-cloudant perspective there is no difference in sending a request to CouchDB or Cloudant Local edition. However, given that this request communicates over the network there are many factors to consider and it is impossible to know which of these might be relevant without understanding your performance testing methodology:

**i) Admin party vs auth:**
Is your CouchDB operating in admin-party mode (i.e. are `COUCH_USERNAME` and `COUCH_PASSWORD` `null` for your CouchDB case)? If so java-cloudant will skip an additional request to `_session` to get and store a cookie. Given that Cloudant Local is almost certainly not operating in admin-party then you might have an extra request round trip on your first request to Cloudant Local compared to CouchDB.

**ii) HTTP vs HTTPS:**
Are both your CouchDB and Cloudant Local operating over `https`? IIRC CouchDB's default installation does not use `https`, but your Cloudant Local installation likely will be. The SSL handshake required for `https` takes longer than making an `http` connection; java-cloudant does pool connections to reduce this impact, but it will be noticeable on the first request made. It is worth noting as well that the [default JVM connection pool purges idle connections fast ~5 s](https://github.com/cloudant/java-cloudant/issues/247#issuecomment-214731668) so `https` connections may need to be renegotiated regularly if you are not making frequent requests.

**iii) First request vs average request time over many requests:**
It is not clear over how many requests you have performed testing, but it is worth noting that the first request made may be slower than subsequent requests because of `i)` and `ii)` above. Further to that if your test is populating the DB each time (as opposed to re-using an existing DB and view) then the first request to the view will cause the view indexes to be built, which will delay the response time. This is true for CouchDB and Cloudant Local, but I guess is more of a point to check that your testing methodology is identical in both cases and because of `iv)` below the time to build the index may be quite different.

**iv) Complexity of Cloudant topology:**
Apache CouchDB 1.x installations are single node and CouchDB 2.x test installations also tend to be single node, whereas a real Cloudant Local installation will have at least 3 nodes. Cloudant Local will have a load balancer proxying those three nodes. Unless you have also configured your CouchDB with 3 nodes and a load balancer you might be comparing apples with oranges.

**v) Network latency/performance**
Do your CouchDB and Cloudant Local installations have identical network performance from your test machine? This will of course impact the time it takes to make the HTTP requests, particularly in the case where points i) or ii) are in play, but also just for transferring the view response data.

If after considering these points you are still concerned about the performance of your Cloudant Local installation then I think you will need to raise it with your Cloudant Local support team.

**Comment by gauravbh1** [comment]
@ricellis Thanks for your reply.
1) I am making the calls to Couch db with credentials.
2) Both dbs are over http.
3) First request is lower for Couch db but subsequent requests are faster but for Cloudant all the calls have the same response time for every call.
I invoked the view in a for loop for 10 times. Here are the details for both

For Couch db 
```
0 = Time diff is 28
1 = Time diff is 3
2 = Time diff is 3
3 = Time diff is 3
4 = Time diff is 3
5 = Time diff is 4
6 = Time diff is 4
7 = Time diff is 3
8 = Time diff is 4
9 = Time diff is 3
```
For Cloudant
```
0 = Time diff is 218
1 = Time diff is 203
2 = Time diff is 210
3 = Time diff is 211
4 = Time diff is 210
5 = Time diff is 210
6 = Time diff is 210
7 = Time diff is 210
8 = Time diff is 211
9 = Time diff is 200
```
4) Your point#4 is worth looking at. I will try to find out what the difference our Couch DB has as compared to Cloudant.

5) Both of these dbs are in local machine and for accessing the sample view, the response times through Fauxton UI are identical.

**Comment by ykodgule** [comment]
_Your point#4 is worth looking at. I will try to find out what the difference our Couch DB has as compared to Cloudant._

Gaurav,
For now, we are using a single node deployment only. So we will have to strike out this possibility as well.
We followed this document for the same:
https://www.ibm.com/support/knowledgecenter/SSTPQH_1.1.0/com.ibm.cloudant.local.install.doc/topics/clinstall_install_single_node.html

**Comment by ricellis** [comment]
As I said previously the requests made are identical from a java-cloudant perspective, excepting the auth scenarios I already outlined that you have now ruled out. As such I don't believe that this is a client side issue.

Just in case I've reviewed some of our CI test timings for the `queryView` test, which is not a performance test, but can at least highlight differences between destinations:

| | apache/couchdb:1.7.1 | apache/couchdb:2.1.1 | ibmcom/cloudant-developer:latest (Cloudant Local 1.1.0)
| --- | --- | --- | --- 
| Local docker container | 79 ms | 168 ms | 175 ms
| Remote docker container | 2.725 s | 3.25 s | 2.975 s

As you can see the couchdb:2.1.1 timings are very similar to that for Cloudant Local and using a real network makes a massive difference, but I can't see the same issue.

One other suggestion would be to run something like `curl -w \\n%{time_total}\\n -u COUCH_USERNAME:COUCH_PASSWORD http://COUCHDB_URL/mydb1/_design/mydesigndoc/_view/testView` (i.e. curl the same view URL) to try and demonstrate that the timings are not different outside of java-cloudant. It's worth keeping in mind that the block you are timing in java-cloudant does include deserializing the response into Java objects (including meta-data returned with the view) so it may also be worth comparing the view response returned from each server as well to validate they are the same.

I don't think there is much else I can do with this issue as I'm unable to help you with specifics or tuning of your Cloudant Local/Couch installations.

**Comment by ykodgule** [comment]
Hi @ricellis ,
Thank you for your responses so far.
We did some more tests on this and see that the huge differences in response times are only seen with the Java cloudant apis. If we use a third party HTTPUtils to retrieve the same docs via the same view definition over REST invocations, then the responses in both the cases (cloudant-local vs couchdb) are more or less the same.
However, every time we test the Java API invocation with couch the responses are always less than 10 milisec. While with Cloudant-local they are always greater than 50milisec. 
even if we consider the time delays with Java APIs due to serialization and deserialization, it should be consistently same for Couch as well as Cloudant-Local. But we dont find it to be the same.
What could be the reason?
Again, if there is any performance tuning guide that can help us?

**Comment by smithsz** [comment]
The library code path for CouchDB and Cloudant-local will be identical. There's certainly nothing from within the library that we do differently that would explain the variations you're seeing.
Have you tried attaching a profiler to establish where the additional time is being spent?
I'm reluctant to believe this is actually a library issue. It might be worth reviewing your Cloudant-local setup. The [support team](mailto:support@cloudant.com) will be able to advise on your current configuration.

**Comment by gauravbh1** [comment]
@ricellis We will try to do some profiling and get back to you with the results. One more thing that I wanted to mention was that the curl command that you provided also resulted in almost similar times for both couch and cloudant local docker containers (which was pretty less). 

Also if we loop around and call a view using the cloudant java api, for couch, the response times get lower after the 1st call. So e.g. for couch the response times are 25ms for the first call and for the subsequent calls to the same view through java api, the response time is around 3ms. However, for cloudant local, the response times are always almost the same. In our case, it's close to 60ms.  As I mentioned, we will try to do some profiling. But if possible can you check at your end the response times for both couch and cloudant view invocation through cloudant java api when invoked in let's say a loop of 10?

**Comment by ricellis** [comment]
Using the same view and 3 repeated attempts of looping 10 requests for each of Cloudant Local and CouchDB 2.1.1 with java-cloudant 2.9.0 (to match your version):

Endpoint |  1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ----
Cloudant Local | 106 | 14 | 16 | 21 | 16 | 12 | 13 | 11 | 14 | 12
Cloudant Local | 78| 14| 15| 18| 21| 15| 22| 12| 11| 12
Cloudant Local | 141| 28| 14| 28| 22| 18| 14| 12| 11| 11
CouchDB 2.1.1 | 108| 9| 8| 7| 7| 7| 7| 7| 9| 10
CouchDB 2.1.1 | 108| 9| 8| 7| 8| 7| 8| 9| 8| 13
CouchDB 2.1.1 | 120| 17| 9| 8| 17| 10| 7| 7| 8| 11

For comparison the same view using curl:

Endpoint |  1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10
---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ---- | ----
Cloudant Local | 118 | 23 | 21 | 23 | 35 | 18 | 16 | 30 | 23 | 14
CouchDB 2.1.1 | 113 | 14 | 33 | 12 | 12 | 12 | 13 | 28 | 12 | 14

In conclusion the java-cloudant API does not make a difference here with the default settings.

**Comment by tomblench** [comment]
Closing old ticket, please re-open if you still have issues.
